sssd_test_framework/utils/sbus.py

```diff
@@ -0,0 +1,672 @@
+        for k, v in self.value.items():
+            val += f"{delim}{str(k)},{str(v)}"
+            delim = ","
+        return val
+
+    @classmethod
+    def _parse_dict_entry(cls, result: DBUSResult) -> tuple:
+        res = re.match(r"^ *dict entry *\(", result.text)
+        if res is None:
+            raise TypeError("Not a dictionary entry")
+
+        result.text = result.text.removeprefix(res.group(0))
+        key = cls.get_python_value(result)
+        value = cls.get_python_value(result)
+
+        res = re.match(r"^ *\) *", result.text)
+        if res is None:
+            raise TypeError("Malformed dictionary entry")
+
+        result.text = result.text.removeprefix(res.group(0)).strip()
+        return key, value
+
+    @classmethod
+    def parse(cls, result: DBUSResult) -> DBUSTypeDict:
+        res = re.match(r"^ *array +\[ *(dict entry *\(.+\)) *\] *", result.text)
+        if res is None:
+            raise TypeError("Not a dictionary")
+
+        res2 = DBUSResult(res.group(1).strip())
+        dict = {}
+        while len(res2.text) > 0:
+            key, value = cls._parse_dict_entry(res2)
+            dict[key] = value.value if isinstance(value, DBUSTypeVariant) else value
+
+        result.text = result.text[res.span()[1] :].strip()
+        return DBUSTypeDict(dict)
+
+
+class DBUSObject(MultihostUtility[MultihostHost]):
+    """
+    D-Bus access.
+
+    Objects of this class represent the root or child D-Bus nodes.
+    """
+
+    class ProxyObject(ABC):
+        """
+        Abstract class representing a D-Bus object (method or property).
+
+        Corresponding concrete classes are provided for their instantiation.
+        """
+
+        def __init__(
+            self,
+            host: MultihostHost,
+            dest: str,
+            objpath: str,
+            bus: str | DBUSBus,
+            /,
+            interface: str | None = None,
+            child: xml.Element | None = None,
+        ):
+            """
+            Constructor to be called by the subclasses either explicitely or by inheritance.
+            """
+            match bus:
+                case DBUSBus.SYSTEM:
+                    _bus_param = "--system"
+                case DBUSBus.SESSION:
+                    _bus_param = "--session"
+                case _:
+                    _bus_param = "--bus=" + bus
+
+            self.host = host
+            self.dest = dest
+            self.objpath = objpath
+            self.bus = bus
+            self.interface = objpath[1:].replace("/", ".") if interface is None else interface
+            self.child = child
+            self.cmd = f"dbus-send {_bus_param} --print-reply --dest={dest} {self.objpath}"
+
+        def _run(self, command: str, *args):
+            """
+            Run dbus-send over ssh on the remote host.
+            """
+            cmd = self.cmd + f" {command}"
+            for param in args:
+                cmd += " " + DBUSType.get_dbus_param(param)
```

Comment:
The parameter type information is present in the introspection, we need to make sure we send what is expected. This is especially important for numbers since it is not possible to differentiate between uint32, uint64, ...

Same for variant, you can't deduce that you want to pass string as variant.